etna/transforms/decomposition/model_based.py

```diff
@@ -0,0 +1,185 @@
+from typing import List
+from typing import Union
+from typing import get_args
+
+import pandas as pd
+
+from etna.datasets import TSDataset
+from etna.datasets.utils import determine_num_steps
+from etna.models import BATSModel
+from etna.models import DeadlineMovingAverageModel
+from etna.models import HoltWintersModel
+from etna.models import ProphetModel
+from etna.models import SARIMAXModel
+from etna.models import SeasonalMovingAverageModel
+from etna.models import TBATSModel
+from etna.models.base import ContextRequiredModelType
+from etna.models.base import ModelType
+from etna.transforms import IrreversibleTransform
+
+_SUPPORTED_MODELS = Union[
+    HoltWintersModel,  # full
+    ProphetModel,  # full
+    SARIMAXModel,  # full
+    DeadlineMovingAverageModel,  # need to account context/prediction size
+    SeasonalMovingAverageModel,  # need to account context/prediction size
+    BATSModel,  # dynamic components, not reliable
+    TBATSModel,  # dynamic components, not reliable
+]
+
+
+class ModelDecomposeTransform(IrreversibleTransform):
+    """Transform that uses ETNA models to estimate series decomposition.
+
+    Note
+    ----
+    This transform decomposes only in-sample data. For the future timestamps it produces ``NaN``.
+    For the dataset to be transformed, it should contain at least the minimum amount of in-sample timestamps that are required by the model.
+    """
+
+    def __init__(self, model: ModelType, in_column: str = "target", residuals: bool = False):
+        """Init ``ModelDecomposeTransform``.
+
+        Parameters
+        ----------
+        model:
+            instance of the model to use for the decomposition. Note that not all models are supported. Possible selections are:
+
+            - ``HoltWintersModel``
+            - ``ProphetModel``
+            - ``SARIMAXModel``
+            - ``DeadlineMovingAverageModel``
+            - ``SeasonalMovingAverageModel``
+            - ``BATSModel``
```

Comment:
We should probably explain possible problems with this model here.

Reply:
Added info